src/MessageHandler/ActivityPub/Inbox/ChainActivityHandler.php

```diff
@@ -35,121 +33,83 @@ public function __construct(
         }
     }
 
-    private function unloadStack(array $chain, array $parent, array $announce = null, array $like = null, array $dislike = null): void
+    private function retrieveObject(string $apUrl): void
     {
-        $object = end($chain);
-
-        if (!empty($object)) {
-            if (\is_string($object)) {
-                if (false !== filter_var($object, FILTER_VALIDATE_URL)) {
-                    $object = $this->client->getActivityObject($object);
-                } else {
-                    // if the object is a string, but not an url it is not valid and therefore nothing should be done
-                    return;
-                }
-            }
-            $createdObject = match ($this->getType($object)) {
-                'Question', 'Note' => $this->note->create($object),
-                'Page' => $this->page->create($object),
-                default => null
-            };
-
-            if ($createdObject and ($createdObject instanceof EntryComment or $createdObject instanceof Post or $createdObject instanceof PostComment)) {
-                if (null !== $createdObject->apId and null === $createdObject->magazine->apId) {
-                    // local magazine, but remote post
-                    $this->bus->dispatch(new OutboxAnnounceMessage(null, $createdObject->magazine->getId(), $createdObject->getId(), \get_class($createdObject)));
-                }
-            }
-
-            array_pop($chain);
-
-            if (\count(array_filter($chain))) {
-                $this->bus->dispatch(new ChainActivityMessage($chain, $parent, $announce, $like, $dislike));
+        try {
+            $object = $this->client->getActivityObject($apUrl);
+            if (!$object) {
+                $this->logger->warning('Got an empty object for {url}', ['url' => $apUrl]);
 
                 return;
             }
-        }
-        // only one of $announce, $like and $dislike should ever be set
-        if ($announce) {
-            $this->bus->dispatch(new AnnounceMessage($announce));
-
-            return;
-        }
+            if (!\is_array($object)) {
+                $this->logger->warning("Didn't get an array for {url}. Got '{val}' instead, exiting", ['url' => $apUrl, 'val' => $object]);
 
-        if ($like) {
-            $this->bus->dispatch(new LikeMessage($like));
-
-            return;
-        }
+                return;
+            }
 
-        if ($dislike) {
-            $this->bus->dispatch(new DislikeMessage($dislike));
+            if (\array_key_exists('inReplyTo', $object) && null !== $object['inReplyTo']) {
+                $parentUrl = \is_string($object['inReplyTo']) ? $object['inReplyTo'] : $object['inReplyTo']['id'];
+                $meta = $this->repository->findByObjectId($parentUrl);
+                if (!$meta) {
+                    $this->retrieveObject($parentUrl);
+                }
+                $meta = $this->repository->findByObjectId($parentUrl);
```

Comment:
Is this intentionally? You are doing the same call twice now?

Reply:
yes it is intentionally. It checks whether the object exists, if it doesn't is calls a method that should create it and afterwards it checks if it actually did create it. Maybe I could make `retrieveObject` return the object instead then I wouldn't need the second db call...

Reply:
`retrieve` it's in the name yes 👍🏽

Reply:
Shouldn't be blocking for now.